chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,409 @@
+      this_coin_bundle         ;; verified with ASSERT_MY_COIN_ID
+      next_coin_bundle         ;; used to generate ASSERT_COIN_ANNOUNCEMENT
+      prev_subtotal            ;; included in announcement, prev_coin ASSERT_COIN_ANNOUNCEMENT will fail if wrong
+      extra_delta              ;; this is the number of "legal discrepancy" between your real delta and what you're announcing your delta is
+    )
+
+     ;;;;; start library code
+
+     (include condition_codes.clvm)
+     (include curry-and-treehash.clinc)
+     (include cat_truths.clib)
+
+     (defmacro assert items
+         (if (r items)
+             (list if (f items) (c assert (r items)) (q . (x)))
+             (f items)
+         )
+     )
+
+     (defun-inline not (arg)
+      (i arg 0 1)
+     )
+
+     ; takes a lisp tree and returns the hash of it
+     (defun sha256tree1 (TREE)
+           (if (l TREE)
+               (sha256 2 (sha256tree1 (f TREE)) (sha256tree1 (r TREE)))
+               (sha256 1 TREE)))
+
+    ; cat_mod_struct = (MOD_HASH MOD_HASH_hash GENESIS_COIN_CHECKER GENESIS_COIN_CHECKER_hash)
+
+    (defun-inline mod_hash_from_cat_mod_struct (cat_mod_struct) (f cat_mod_struct))
+    (defun-inline mod_hash_hash_from_cat_mod_struct (cat_mod_struct) (f (r cat_mod_struct)))
+    (defun-inline genesis_coin_checker_hash_from_cat_mod_struct (cat_mod_struct) (f (r (r (r cat_mod_struct)))))
+
+     ;;;;; end library code
+
+     ;; return the puzzle hash for a cat with the given `GENESIS_COIN_CHECKER_hash` & `INNER_PUZZLE`
+     (defun-inline cat_puzzle_hash (cat_mod_struct inner_puzzle_hash)
+       (puzzle-hash-of-curried-function (mod_hash_from_cat_mod_struct cat_mod_struct)
+                                        inner_puzzle_hash
+                                        (sha256 ONE (genesis_coin_checker_hash_from_cat_mod_struct cat_mod_struct))
+                                        (mod_hash_hash_from_cat_mod_struct cat_mod_struct)
+       )
+     )
+
+     ;; tweak `CREATE_COIN` condition by wrapping the puzzle hash, forcing it to be a cat
+     ;; prohibit CREATE_COIN_ANNOUNCEMENT
+
+     (defun-inline morph_condition (condition cat_mod_struct)
+       (if (= (f condition) CREATE_COIN)
+         (c CREATE_COIN
+               (c (cat_puzzle_hash cat_mod_struct (f (r condition)))
+                    (r (r condition)))
+         )
+         (if (= (f condition) CREATE_COIN_ANNOUNCEMENT)
+           (list CREATE_COIN_ANNOUNCEMENT (concat 0xca (f (r condition))))
+           condition
+         )
+       )
+     )
+
+     ;; tweak all `CREATE_COIN` conditions, enforcing created coins to be cats
+     (defun morph_conditions (conditions cat_mod_struct)
+       (if conditions
+         (c
+           (morph_condition (f conditions) cat_mod_struct)
+           (morph_conditions (r conditions) cat_mod_struct)
+         )
+         ()
+       )
+     )
+
+     ;; given a coin triplet, return the id of the coin
+     (defun-inline coin_id_for_coin (coin cat_mod_struct)
```

Comment:
What does this do now? The comment is wrong and I can't figure out what's supposed to be happening here.

Reply:
It calculates a coin id, but first wraps the puzzle hash in a CAT mod.  This is how we assert that our neighbor is a CAT.